File: ffi/src/server_message.rs
use crate::{
    directory::Directory,
    error::{ErrorCode, Result},
    network::NetworkEvent,
    registry::Handle,
};
use ouisync_lib::{PeerInfo, Progress, StateMonitor};
use serde::Serialize;
use serde_bytes::ByteBuf;
use std::net::SocketAddr;

#[derive(Serialize)]
#[serde(rename_all = "snake_case")]
pub(crate) enum ServerMessage {
    Success(Value),
    Failure { code: ErrorCode, message: String },
    Notification(Notification),
}

impl ServerMessage {
    pub fn response(result: Result<Value>) -> Self {
        match result {
            Ok(response) => Self::Success(response),
            Err(error) => Self::Failure {
                code: error.to_error_code(),
                message: error.to_string(),
            },
        }
    }

    pub fn notification(notification: Notification) -> Self {
        Self::Notification(notification)
    }
}

#[derive(Serialize)]
#[serde(untagged)]
pub(crate) enum Value {
    None,
    Bool(bool),
    U8(u8),
    U32(u32),
    U64(u64),
    Bytes(ByteBuf),
    String(String),
    Handle(u64),
    Directory(Directory),
    StateMonitor(StateMonitor),
    Progress(Progress),
    PeerInfo(Vec<PeerInfo>),
}

impl<T> From<Option<T>> for Value
where
    Value: From<T>,
{
    fn from(value: Option<T>) -> Self {
        if let Some(value) = value {
            Self::from(value)
        } else {
            Self::None
        }
    }
}

impl From<()> for Value {
    fn from(_: ()) -> Self {
        Self::None
    }
}

impl From<bool> for Value {
    fn from(value: bool) -> Self {
        Self::Bool(value)
    }
}

impl From<u8> for Value {
    fn from(value: u8) -> Self {
        Self::U8(value)
    }
}

impl From<u32> for Value {
    fn from(value: u32) -> Self {
        Self::U32(value)
    }
}

impl From<u64> for Value {
    fn from(value: u64) -> Self {
        Self::U64(value)
    }
}

impl From<Vec<u8>> for Value {
    fn from(value: Vec<u8>) -> Self {
        Self::Bytes(ByteBuf::from(value))
    }
}

impl From<String> for Value {
    fn from(value: String) -> Self {
        Self::String(value)
    }
}

impl From<StateMonitor> for Value {
    fn from(value: StateMonitor) -> Self {
        Self::StateMonitor(value)
    }
}

impl From<Directory> for Value {
    fn from(value: Directory) -> Self {
        Self::Directory(value)
    }
}

impl<T> From<Handle<T>> for Value {
    fn from(value: Handle<T>) -> Self {
        Self::Handle(value.id())
    }
}

impl From<SocketAddr> for Value {
    fn from(value: SocketAddr) -> Self {
        Self::String(value.to_string())
    }
}

impl From<Progress> for Value {
    fn from(value: Progress) -> Self {
        Self::Progress(value)
    }
}

impl From<Vec<PeerInfo>> for Value {
    fn from(value: Vec<PeerInfo>) -> Self {
        Self::PeerInfo(value)
    }
}

#[derive(Serialize)]
#[serde(untagged)]
pub(crate) enum Notification {
    Repository,
    Network(NetworkEvent),
    StateMonitor,
}

File: bindings/kotlin/src/main/kotlin/org/equalitie/ouisync/ServerMessage.kt
package org.equalitie.ouisync

import org.msgpack.core.MessageUnpacker
import org.msgpack.value.ValueType

data class PeerInfo(
    val ip: String,
    val port: UShort,
    val source: String,
    val state: String,
    val runtimeId: String?,
) {
    companion object {
        fun unpack(unpacker: MessageUnpacker): PeerInfo {
            val count = unpacker.unpackArrayHeader()

            if (count < 4) {
                throw InvalidResponse()
            }

            val ip = unpacker.unpackString()
            val port = unpacker.unpackInt().toUShort()
            val source = unpacker.unpackString()

            var state: String
            var runtimeId: String? = null

            when (unpacker.getNextFormat().getValueType()) {
                ValueType.STRING -> {
                    state = unpacker.unpackString()
                }
                ValueType.MAP -> {
                    if (unpacker.unpackMapHeader() < 1) {
                        throw InvalidResponse()
                    }

                    state = unpacker.unpackString()

                    val length = unpacker.unpackBinaryHeader()
                    val runtimeIdBytes = unpacker.readPayload(length)

                    @OptIn(kotlin.ExperimentalStdlibApi::class)
                    runtimeId = runtimeIdBytes.toHexString()
                }
                else -> throw InvalidResponse()
            }

            return PeerInfo(ip, port, source, state, runtimeId)
        }
    }
}

enum class NetworkEvent {
    PROTOCOL_VERSION_MISMATCH, // 0
    PEER_SET_CHANGE, // = 1

    ;

    companion object {
        fun fromByte(n: Byte): NetworkEvent = when (n.toInt()) {
            0 -> PROTOCOL_VERSION_MISMATCH
            1 -> PEER_SET_CHANGE
            else -> throw IllegalArgumentException()
        }
    }

    fun toByte(): Byte = when (this) {
        PROTOCOL_VERSION_MISMATCH -> 0
        PEER_SET_CHANGE -> 1
    }
}

internal sealed interface ServerMessage {
    companion object {
        fun unpack(unpacker: MessageUnpacker): ServerMessage {
            if (unpacker.getNextFormat().getValueType() != ValueType.MAP) {
                throw InvalidMessage()
            }

            if (unpacker.unpackMapHeader() < 1) {
                throw InvalidMessage()
            }

            val kind = unpacker.unpackString()

            when (kind) {
                "success" -> return Success.unpack(unpacker)
                "failure" -> return Failure.unpack(unpacker)
                "notification" -> return Notification.unpack(unpacker)
                else -> throw InvalidMessage()
            }
        }
    }
}

internal sealed interface Response : ServerMessage

internal class Success(val value: Any?) : Response {
    companion object {
        fun unpack(unpacker: MessageUnpacker): Response {
            when (unpacker.getNextFormat().getValueType()) {
                ValueType.STRING -> {
                    if (unpacker.unpackString() == "none") {
                        return Success(null)
                    } else {
                        throw InvalidResponse()
                    }
                }
                ValueType.MAP -> {
                    if (unpacker.unpackMapHeader() != 1) {
                        throw InvalidResponse()
                    }

                    val name = unpacker.unpackString()
                    val value = unpackValue(name, unpacker)

                    return Success(value)
                }
                else -> throw InvalidResponse()
            }
        }

        private fun unpackValue(name: String, unpacker: MessageUnpacker): Any? =
            when (unpacker.getNextFormat().getValueType()) {
                ValueType.ARRAY -> {
                    when (name) {
                        "peer_info" -> unpackPeerInfo(unpacker)
                        else -> throw InvalidResponse()
                    }
                }
                ValueType.BOOLEAN -> unpacker.unpackBoolean()
                ValueType.INTEGER -> {
                    when (name) {
                        "u8" -> unpacker.unpackByte()
                        "u32" -> unpacker.unpackInt()
                        "u64", "handle" -> unpacker.unpackLong()
                        else -> throw InvalidResponse()
                    }
                }
                ValueType.STRING -> unpacker.unpackString()
                else -> throw InvalidResponse()
            }

        private fun unpackPeerInfo(unpacker: MessageUnpacker): List<PeerInfo> {
            val count = unpacker.unpackArrayHeader()
            return 0.rangeUntil(count).map { PeerInfo.unpack(unpacker) }
        }
    }
}

internal class Failure(val error: Error) : Response {
    companion object {
        private val INVALID_ERROR = Error(ErrorCode.MALFORMED_MESSAGE, "invalid error response")

        fun unpack(unpacker: MessageUnpacker): Response {
            if (unpacker.getNextFormat().getValueType() != ValueType.ARRAY) {
                return Failure(INVALID_ERROR)
            }

            if (unpacker.unpackArrayHeader() != 2) {
                return Failure(INVALID_ERROR)
            }

            try {
                val code = ErrorCode.fromShort(unpacker.unpackShort())
                val message = unpacker.unpackString()

                return Failure(Error(code, message))
            } catch (e: Exception) {
                return Failure(INVALID_ERROR)
            }
        }
    }
}

internal class Notification(val content: Any?) : ServerMessage {
    companion object {
        fun unpack(unpacker: MessageUnpacker): Notification {
            val name = when (unpacker.getNextFormat().getValueType()) {
                ValueType.STRING -> unpacker.unpackString()
                ValueType.MAP -> {
                    if (unpacker.unpackMapHeader() != 1) {
                        throw InvalidNotification()
                    }

                    unpacker.unpackString()
                }
                else -> throw InvalidNotification()
            }

            val value = unpackValue(name, unpacker)

            return Notification(value)
        }

        private fun unpackValue(name: String, unpacker: MessageUnpacker): Any? {
            when (name) {
                "network" -> return NetworkEvent.fromByte(unpacker.unpackByte())
                else -> throw InvalidNotification()
            }
        }
    }
}

internal open class InvalidMessage : Exception {
    constructor() : super("invalid message")
    constructor(message: String) : super(message)
}

internal class InvalidResponse : InvalidMessage("invalid response")
internal class InvalidNotification : InvalidMessage("invalid notification")

// pub(crate) enum Response {
//     Bytes(#[serde(with = "serde_bytes")] Vec<u8>),
//     Directory(Directory),
//     StateMonitor(StateMonitor),
//     Progress(Progress),
// }
